pkg/reconciler/timeout_handler.go

```diff
@@ -0,0 +1,219 @@
+package reconciler
+
+import (
+	"sync"
+
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	clientset "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	defaultFunc = func(i interface{}, l *zap.SugaredLogger) {}
+)
+
+const (
+	defaultTimeout = 10 * time.Minute
+)
+
+// StatusKey interface to be implemented by Taskrun Pipelinerun types
+type StatusKey interface {
+	GetRunKey() string
+}
+
+// TimeoutSet contains required k8s interfaces to handle build timeouts
+type TimeoutSet struct {
+	logger                  *zap.SugaredLogger
+	kubeclientset           kubernetes.Interface
+	pipelineclientset       clientset.Interface
+	taskRuncallbackFunc     func(interface{})
+	pipelineruncallbackFunc func(interface{})
+	stopCh                  <-chan struct{}
+	statusMap               *sync.Map
+	done                    map[string]chan bool
+	doneMut                 sync.Mutex
+}
+
+// NewTimeoutHandler returns TimeoutSet filled structure
+func NewTimeoutHandler(
+	kubeclientset kubernetes.Interface,
+	pipelineclientset clientset.Interface,
+	stopCh <-chan struct{},
+	logger *zap.SugaredLogger,
+) *TimeoutSet {
+	return &TimeoutSet{
+		kubeclientset:     kubeclientset,
+		pipelineclientset: pipelineclientset,
+		stopCh:            stopCh,
+		statusMap:         &sync.Map{},
+		done:              make(map[string]chan bool),
+		doneMut:           sync.Mutex{},
+		logger:            logger,
+	}
+}
+
+func (t *TimeoutSet) AddTrCallBackFunc(f func(interface{})) {
```

Comment:
This name sounds like it will append a func, when in fact it replaces the func. How about `SetTaskRunCallbackFunc` ?

(and below for PipelineRun)